### PR TITLE
JAVA-2527: Compression

### PR DESCRIPTION
**Pull request by jyemin**
This builds on the refactoring from a previous PR (which hasn't been pushed to master yet, which is why it's the base commit).

Of note:

* It takes an optional dependency on a 3rd party Snappy implementation, so you'll get runtime failures if you select the snappy compressor and it's not included.  This'll have to go in the release notes.  Also, that library is Java 7 only, so another nail in the Java 6 coffin. Another interesting thing about this library is that it uses JNI but bundles its native dependencies in the jar itself, extracting to the file system at runtime. There is a native Snappy implementation available, but it doesn't seem as well used or maintained so I hesitate to depend on it.
* It supports the async path, but it currently assumes that uncompressed messages are replied to with uncompressed messages.  This seems to bear out and is in line with SERVER tickets, but it's technically a violation of the spec, which says "Clients MUST be able to parse both compressed and uncompressed responses to both compressed and uncompressed requests".  But there is no way to test this as no server actually replies with a compressed message to an uncompressed message.
* InternalStreamConnection could use some more unit test coverage.  It's currently thin and doesn't test the compression path.  I'll work on this.

Patch build: https://evergreen.mongodb.com/version/59847f992fbabe5d180003fc

The patch build failures are unrelated to this patch.  They all also have appeared on master.

### CODE REVIEW

File: driver-async/src/main/com/mongodb/async/client/MongoClientSettings.java [hunk 1]
```diff
@@ -317,6 +321,23 @@ public Builder applicationName(final String applicationName) {
...
+         * Sets the compressors to use for compressing messages to the server. The driver will use the first compressor in the list
+         * that the server is configured to support.
+         *
+         * @param compressorList the list of compressors to request
+         * @return {@code this}
+         * @see #getCompressorList() ()
+         * @since 3.6
+         * @mongodb.server.release 3.4
+         */
+
```
Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/MongoClientSettings.java]: nit - newline.

File: driver-core/build.gradle [hunk 2]
```diff
@@ -32,6 +32,7 @@ dependencies {
     compile "io.netty:netty-buffer:$nettyVersion", optional
     compile "io.netty:netty-transport:$nettyVersion", optional
     compile "io.netty:netty-handler:$nettyVersion", optional
+    compile "org.xerial.snappy:snappy-java:1.1.4", optional
```
Comment (rozza) [inline driver-core/build.gradle]: Suggestion - keep the version numbers in one place: `mongo-java-driver/build.gradle`.

File: driver-core/src/main/com/mongodb/ConnectionString.java [hunk 3]
```diff
@@ -181,6 +181,13 @@
  * <li>{@code appName=string}: Sets the logical name of the application.  The application name may be used by the client to identify
  * the application to the server, for use in server logs, slow query logs, and profile collection.</li>
  * </ul>
+ * <p>Compressor configuration:</p>
+ * <ul>
+ * <li>{@code compressors=string}: A comma-separated list of compressors to request from the server.  The only supported compressor
+ * currently is 'zlib'</li>
```
Comment (rozza) [inline driver-core/src/main/com/mongodb/ConnectionString.java]: Mention the snappy support as well?

File: driver-core/src/main/com/mongodb/connection/SnappyCompressor.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    public String getName() {
+        return "snappy";
+    }
+
+    @Override
+    public byte getId() {
+        return 1;
+    }
+
+    // the server does not support the framing format so SnappyFramedOutputStream can't be used.  The entire source message must first
```
Comment (rozza) [inline driver-core/src/main/com/mongodb/connection/SnappyCompressor.java]: Did you add a ticket about framing support? Or is it something that will never be supported?

File: driver/src/main/com/mongodb/MongoClientURI.java [hunk 5]
```diff
@@ -162,6 +162,13 @@
  * <li>{@code appName=string}: Sets the logical name of the application.  The application name may be used by the client to identify
  * the application to the server, for use in server logs, slow query logs, and profile collection.</li>
  * </ul>
+ * <p>Compressor configuration:</p>
+ * <ul>
+ * <li>{@code compressors=string}: A comma-separated list of compressors to request from the server.  The only supported compressor
```
Comment (rozza) [inline driver/src/main/com/mongodb/MongoClientURI.java]: Include snappy support?

File: mongo-java-driver/build.gradle [hunk 6]
```diff
@@ -30,6 +30,7 @@ dependencies {
     compile "io.netty:netty-buffer:$nettyVersion", optional
     compile "io.netty:netty-transport:$nettyVersion", optional
     compile "io.netty:netty-handler:$nettyVersion", optional
+    compile "org.xerial.snappy:snappy-java:1.1.4", optional
```
Comment (rozza) [inline mongo-java-driver/build.gradle]: Suggestion - make the version global

File: driver-core/src/main/com/mongodb/connection/SnappyCompressor.java (hunk 4 above)
Comment (jyemin) [inline driver-core/src/main/com/mongodb/connection/SnappyCompressor.java]: I'm discussing it with the server team.  It will require a new compressor type since framing is not compatible with non-framing.

File: driver-async/src/main/com/mongodb/async/client/MongoClientSettings.java (hunk 1 above)
Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/MongoClientSettings.java]: Done

File: driver-core/build.gradle (hunk 2 above)
Comment (jyemin) [inline driver-core/build.gradle]: Done

File: driver-core/src/main/com/mongodb/ConnectionString.java (hunk 3 above)
Comment (jyemin) [inline driver-core/src/main/com/mongodb/ConnectionString.java]: Done

File: driver/src/main/com/mongodb/MongoClientURI.java (hunk 5 above)
Comment (jyemin) [inline driver/src/main/com/mongodb/MongoClientURI.java]: Done

File: mongo-java-driver/build.gradle (hunk 6 above)
Comment (jyemin) [inline mongo-java-driver/build.gradle]: Done

File: bson/src/main/org/bson/io/BasicOutputBuffer.java [hunk 7]
```diff
@@ -131,7 +132,7 @@ public void truncateToPosition(final int newPosition) {
     @Override
     public List<ByteBuf> getByteBuffers() {
         ensureOpen();
-        return Arrays.<ByteBuf>asList(new ByteBufNIO(ByteBuffer.wrap(buffer, 0, position).duplicate()));
+        return Arrays.<ByteBuf>asList(new ByteBufNIO(ByteBuffer.wrap(buffer, 0, position).duplicate().order(LITTLE_ENDIAN)));
```
Comment (rozza) [inline bson/src/main/org/bson/io/BasicOutputBuffer.java]: Why is this necessary? and could it be a breaking change?

Comment (jyemin) [inline bson/src/main/org/bson/io/BasicOutputBuffer.java]: It's necessary because there's now a consumer that cares about the endianness of the byte buffers.  Previously, all consumers treated this as a raw byte array but now CompressedMessage calls ByteBuf.getInt, which is endian-sensitive.  

While we could change the endian-ness in CompressedMessage, it seems safer to do it in OutputBuffer, since BSON is always little-endian.

As for backwards compatibility, I'm not too concerned, as:

* the endian-ness was previously unspecified
* i can't see why anyone would ever want it to be big-endian, given that BSON is little-endian
* all current known use cases don't care about endian-ness
* it's unlikely anyone else is using this method

**Comment by jyemin** [comment]
I'll add docs here.  Thanks for the reminder.

File: bson/src/main/org/bson/io/BasicOutputBuffer.java (hunk 7 above)
Comment (rozza) [inline bson/src/main/org/bson/io/BasicOutputBuffer.java]: 👍  thanks

**Comment by jyemin** [comment]
* Added a new unit test for InternalStreamConnection
* Added reference documentation

File: docs/reference/content/driver-async/tutorials/compression.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+MongoClient client = MongoClients.create(settings);
+```
+
+to configure multiple compressors. 
+
+As with configuration with a URI, the driver will use the first compressor in the list for which the server advertises support. 
+
+### Dependencies
+
+As the JDK has no built-in support for Snappy, the driver takes a dependency on an existing open-source Snappy implementation.  See the
```
Comment (rozza) [inline docs/reference/content/driver-async/tutorials/compression.md]: Should we explicitly state the version supported?
